File: core/src/main/java/com/jetbrains/tmp/learning/StepikProjectManager.java
package com.jetbrains.tmp.learning;

import com.google.gson.internal.LinkedTreeMap;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.components.PersistentStateComponent;
import com.intellij.openapi.components.ServiceManager;
import com.intellij.openapi.components.State;
import com.intellij.openapi.components.Storage;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.module.ModifiableModuleModel;
import com.intellij.openapi.module.ModuleManager;
import com.intellij.openapi.module.ModuleWithNameAlreadyExists;
import com.intellij.openapi.options.ConfigurationException;
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.progress.ProgressManager;
import com.intellij.openapi.project.DumbAware;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.vfs.VirtualFileManager;
import com.jetbrains.tmp.learning.core.EduNames;
import com.jetbrains.tmp.learning.courseFormat.CourseNode;
import com.jetbrains.tmp.learning.courseFormat.LessonNode;
import com.jetbrains.tmp.learning.courseFormat.SectionNode;
import com.jetbrains.tmp.learning.courseFormat.StepNode;
import com.jetbrains.tmp.learning.courseFormat.StudyNode;
import com.jetbrains.tmp.learning.serialization.SampleConverter;
import com.jetbrains.tmp.learning.serialization.StudySerializationUtils;
import com.jetbrains.tmp.learning.serialization.StudyUnrecognizedFormatException;
import com.jetbrains.tmp.learning.serialization.SupportedLanguagesConverter;
import com.jetbrains.tmp.learning.ui.StudyToolWindow;
import com.thoughtworks.xstream.XStream;
import com.thoughtworks.xstream.XStreamException;
import com.thoughtworks.xstream.annotations.XStreamOmitField;
import com.thoughtworks.xstream.io.xml.DomDriver;
import org.jdom.Element;
import org.jdom.JDOMException;
import org.jdom.input.DOMBuilder;
import org.jdom.output.XMLOutputter;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.stepik.api.objects.courses.Course;
import org.stepik.api.objects.lessons.CompoundUnitLesson;
import org.stepik.api.objects.sections.Section;
import org.stepik.api.objects.steps.Limit;
import org.stepik.api.objects.steps.Sample;
import org.stepik.api.objects.steps.Step;
import org.stepik.api.objects.steps.VideoUrl;
import org.stepik.api.objects.users.User;
import org.stepik.plugin.projectWizard.idea.SandboxModuleBuilder;
import org.w3c.dom.Document;
import org.xml.sax.SAXException;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.util.UUID;

import static com.jetbrains.tmp.learning.SupportedLanguages.INVALID;
import static java.nio.charset.StandardCharsets.UTF_8;
import static org.stepik.core.utils.ProjectFilesUtils.getOrCreateSrcDirectory;

@State(name = "StepikStudySettings", storages = @Storage("stepik_study_project.xml"))
public class StepikProjectManager implements PersistentStateComponent<Element>, DumbAware {
    private static final int CURRENT_VERSION = 4;
    private static final Logger logger = Logger.getInstance(StepikProjectManager.class);
    @XStreamOmitField
    private static XStream xStream;
    @XStreamOmitField
    private static XMLOutputter outputter;
    @XStreamOmitField
    private static DocumentBuilderFactory factory;
    @XStreamOmitField
    private static DocumentBuilder builder;
    @XStreamOmitField
    private static DOMBuilder domBuilder;
    @XStreamOmitField
    private final Project project;
    private StudyNode<?, ?> root;
    private StudyNode<?, ?> selected;
    private boolean showHint = false;
    private long createdBy;
    private SupportedLanguages defaultLang = INVALID;
    private int version = CURRENT_VERSION;
    private String uuid;

    public StepikProjectManager(@Nullable Project project) {
        this.project = project;
    }

    public StepikProjectManager() {
        this(null);
    }

    @Nullable
    public static StepikProjectManager getInstance(@NotNull final Project project) {
        return ServiceManager.getService(project, StepikProjectManager.class);
    }

    @Nullable
    public static StudyNode getProjectRoot(@NotNull final Project project) {
        StepikProjectManager instance = getInstance(project);

        if (instance == null) {
            return null;
        }

        return instance.getProjectRoot();
    }

    public static boolean isStepikProject(@Nullable Project project) {
        if (project == null) {
            return false;
        }
        StepikProjectManager instance = getInstance(project);
        return instance != null && instance.getProjectRoot() != null;
    }

    private static String elementToXml(@NotNull Element state) {
        if (outputter == null) {
            outputter = new XMLOutputter();
        }
        return outputter.outputString(state.getChild(StudySerializationUtils.MAIN_ELEMENT));
    }

    @NotNull
    public static XStream getXStream() {
        if (xStream == null) {
            xStream = new XStream(new DomDriver());
            xStream.alias("StepikProjectManager", StepikProjectManager.class);
            xStream.alias("CourseNode", CourseNode.class);
            xStream.alias("SectionNode", SectionNode.class);
            xStream.alias("LessonNode", LessonNode.class);
            xStream.alias("StepNode", StepNode.class);
            xStream.alias("Limit", Limit.class);
            xStream.alias("SupportedLanguages", SupportedLanguages.class);
            xStream.alias("VideoUrl", VideoUrl.class);
            xStream.alias("LinkedTreeMap", LinkedTreeMap.class);
            xStream.alias("Sample", Sample.class);
            xStream.alias("Course", Course.class);
            xStream.alias("Section", Section.class);
            xStream.alias("CompoundUnitLesson", CompoundUnitLesson.class);
            xStream.alias("Step", Step.class);
            xStream.alias("User", User.class);
            xStream.autodetectAnnotations(true);
            xStream.setClassLoader(StepikProjectManager.class.getClassLoader());
            xStream.registerConverter(new SupportedLanguagesConverter());
            xStream.registerConverter(new SampleConverter());
            xStream.ignoreUnknownElements();
            xStream.setMode(XStream.ID_REFERENCES);
        }

        return xStream;
    }

    @NotNull
    private static Element toElement(ByteArrayOutputStream out)
            throws ParserConfigurationException, SAXException, IOException {
        if (factory == null) {
            factory = DocumentBuilderFactory.newInstance();
            factory.setValidating(false);
            builder = factory.newDocumentBuilder();
            domBuilder = new DOMBuilder();
        }
        try (ByteArrayInputStream in = new ByteArrayInputStream(out.toByteArray())) {
            Document doc = builder.parse(in);
            org.jdom.Document document = domBuilder.build(doc);

            Element root = document.getRootElement();
            document.removeContent(root);

            Element element = new Element("element");
            element.addContent(root);
            return element;
        }
    }

    public static void setSelected(@NotNull Project project, @Nullable StudyNode studyNode, boolean force) {
        StepikProjectManager instance = getInstance(project);
        if (instance != null) {
            instance.setSelected(studyNode, force);
        }
    }

    public static void setSelected(@NotNull Project project, @Nullable StudyNode studyNode) {
        setSelected(project, studyNode, false);
    }

    public static void updateSelection(@NotNull Project project) {
        StepikProjectManager instance = getInstance(project);
        if (instance != null) {
            instance.updateSelection();
        }
    }

    @Nullable
    public static StudyNode<?, ?> getSelected(@Nullable Project project) {
        if (project == null) {
            return null;
        }
        StepikProjectManager instance = getInstance(project);
        if (instance != null) {
            return instance.getSelected();
        }

        return null;
    }

    public StudyNode<?, ?> getSelected() {
        return selected;
    }

    public void setSelected(StudyNode<?, ?> selected) {
        setSelected(selected, false);
    }

    public void setSelected(StudyNode<?, ?> selected, boolean force) {
        this.selected = selected;
        if (project != null) {
            StudyToolWindow toolWindow = StudyUtils.getStudyToolWindow(project);
            if (toolWindow != null) {
                ApplicationManager.getApplication().invokeLater(() -> toolWindow.setStepNode(selected, force));
            }
        }
    }

    public void updateSelection() {
        setSelected(selected, true);
    }

    public void setRootNode(@Nullable StudyNode root) {
        this.root = root;
    }

    @Nullable
    @Override
    public Element getState() {
        if (getProjectRoot() == null) {
            return null;
        }
        try (ByteArrayOutputStream out = new ByteArrayOutputStream()) {
            getXStream().toXML(this, new OutputStreamWriter(out, UTF_8));
            Element el = toElement(out);
            logger.info("Getting the StepikProjectManager state");

            return el;
        } catch (ParserConfigurationException | IOException | SAXException e) {
            logger.warn("Failed getting the StepikProjectManager state", e);
        }

        return null;
    }

    @Override
    public void loadState(Element state) {
        try {
            logger.info("Start load the StepikProjectManager state");
            int version = StudySerializationUtils.getVersion(state);

            switch (version) {
                case 1:
                    state = StudySerializationUtils.convertToSecondVersion(state);
                case 2:
                    state = StudySerializationUtils.convertToThirdVersion(state);
                case 3:
                    state = StudySerializationUtils.convertToFourthVersion(state);
                    //uncomment for future versions
                    //case 4:
                    //state = StudySerializationUtils.convertToFifthVersion(state);
            }

            String xml = elementToXml(state);
            XStream xs = getXStream();
            xs.fromXML(xml, this);

            this.version = CURRENT_VERSION;
            refreshCourse();
            logger.info("The StepikProjectManager state loaded");
        } catch (XStreamException | StudyUnrecognizedFormatException e) {
            logger.warn("Failed deserialization StepikProjectManager \n" + e.getMessage() + "\n" + project);
        }
    }

    private void refreshCourse() {
        if (getProjectRoot() == null) {
            return;
        }

        ProgressManager.getInstance()
                .runProcessWithProgressSynchronously(() -> {
                    ProgressIndicator indicator = ProgressManager.getInstance()
                            .getProgressIndicator();
                    indicator.setIndeterminate(true);
                    root.reloadData(indicator);
                }, "Refreshing Course", true, project);
        ApplicationManager.getApplication().invokeLater(() -> {
            repairProjectFiles(root);
            VirtualFileManager.getInstance().syncRefresh();

            VirtualFile projectDir = project != null ? project.getBaseDir() : null;
            if (projectDir != null && projectDir.findChild(EduNames.SANDBOX_DIR) == null) {
                ModifiableModuleModel model = ModuleManager.getInstance(project).getModifiableModel();

                ApplicationManager.getApplication().runWriteAction(() -> {
                    try {
                        new SandboxModuleBuilder(projectDir.getPath()).createModule(model);
                        model.commit();
                    } catch (IOException | ConfigurationException | JDOMException | ModuleWithNameAlreadyExists e) {
                        logger.warn("Failed repair Sandbox", e);
                    }
                });
            }
        });
    }

    private void repairProjectFiles(@NotNull StudyNode<?, ?> node) {
        if (project != null) {
            if (node instanceof StepNode) {
                getOrCreateSrcDirectory(project, (StepNode) node, false);
            }
            node.getChildren().forEach(this::repairProjectFiles);
        }
    }

    @NotNull
    public SupportedLanguages getDefaultLang() {
        return defaultLang;
    }

    public void setDefaultLang(@NotNull SupportedLanguages defaultLang) {
        this.defaultLang = defaultLang;
    }

    public boolean getShowHint() {
        return showHint;
    }

    public void setShowHint(boolean showHint) {
        this.showHint = showHint;
    }

    public Project getProject() {
        return project;
    }

    public int getVersion() {
        return version;
    }

    public void setCreatedBy(long createdBy) {
        this.createdBy = createdBy;
    }

    @NotNull
    public String getUuid() {
        if (uuid == null) {
            uuid = UUID.randomUUID().toString();
        }
        return uuid;
    }

    @Nullable
    public StudyNode getProjectRoot() {
        return root;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        StepikProjectManager that = (StepikProjectManager) o;

        if (showHint != that.showHint) return false;
        if (createdBy != that.createdBy) return false;
        if (version != that.version) return false;
        if (project != null ? !project.equals(that.project) : that.project != null) return false;
        if (root != null ? !root.equals(that.root) : that.root != null) return false;
        //noinspection SimplifiableIfStatement
        if (defaultLang != that.defaultLang) return false;
        return uuid != null ? uuid.equals(that.uuid) : that.uuid == null;
    }

    @Override
    public int hashCode() {
        int result = project != null ? project.hashCode() : 0;
        result = 31 * result + (root != null ? root.hashCode() : 0);
        result = 31 * result + (showHint ? 1 : 0);
        result = 31 * result + (int) (createdBy ^ (createdBy >>> 32));
        result = 31 * result + (defaultLang != null ? defaultLang.hashCode() : 0);
        result = 31 * result + version;
        result = 31 * result + (uuid != null ? uuid.hashCode() : 0);
        return result;
    }
}

File: stepik-union/src/main/java/org/stepik/core/StepikProjectManager.kt
package org.stepik.core

import com.google.gson.internal.LinkedTreeMap
import com.intellij.ide.projectView.ProjectView
import com.intellij.openapi.Disposable
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.components.PersistentStateComponent
import com.intellij.openapi.components.ServiceManager
import com.intellij.openapi.components.State
import com.intellij.openapi.components.Storage
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.fileEditor.FileEditorManager
import com.intellij.openapi.module.ModuleManager
import com.intellij.openapi.module.ModuleWithNameAlreadyExists
import com.intellij.openapi.options.ConfigurationException
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.progress.Task
import com.intellij.openapi.project.DumbAware
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.Computable
import com.intellij.openapi.util.Disposer
import com.intellij.openapi.vfs.VirtualFileManager
import com.thoughtworks.xstream.XStream
import com.thoughtworks.xstream.XStreamException
import com.thoughtworks.xstream.annotations.XStreamOmitField
import com.thoughtworks.xstream.io.xml.DomDriver
import org.jdom.Element
import org.jdom.JDOMException
import org.jdom.input.DOMBuilder
import org.jdom.output.XMLOutputter
import org.stepik.api.objects.courses.Course
import org.stepik.api.objects.lessons.CompoundUnitLesson
import org.stepik.api.objects.sections.Section
import org.stepik.api.objects.steps.Limit
import org.stepik.api.objects.steps.Sample
import org.stepik.api.objects.steps.Step
import org.stepik.api.objects.steps.VideoUrl
import org.stepik.api.objects.users.User
import org.stepik.core.core.EduNames
import org.stepik.core.courseFormat.CourseNode
import org.stepik.core.courseFormat.LessonNode
import org.stepik.core.courseFormat.SectionNode
import org.stepik.core.courseFormat.StepNode
import org.stepik.core.courseFormat.StudyNode
import org.stepik.core.projectWizard.idea.SandboxModuleBuilder
import org.stepik.core.serialization.SampleConverter
import org.stepik.core.serialization.StudySerializationUtils
import org.stepik.core.serialization.StudyUnrecognizedFormatException
import org.stepik.core.serialization.SupportedLanguagesConverter
import org.stepik.core.stepik.StepikAuthManager
import org.stepik.core.stepik.StepikAuthManager.authAndGetStepikApiClient
import org.stepik.core.stepik.StepikAuthManager.isAuthenticated
import org.stepik.core.stepik.StepikAuthManagerListener
import org.stepik.core.stepik.StepikAuthState
import org.stepik.core.stepik.StepikAuthState.AUTH
import org.stepik.core.stepik.StepikAuthState.NOT_AUTH
import org.stepik.core.stepik.StepikAuthState.SHOW_DIALOG
import org.stepik.core.utils.ProjectFilesUtils
import org.xml.sax.SAXException
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.OutputStreamWriter
import java.nio.charset.StandardCharsets.UTF_8
import java.util.*
import java.util.concurrent.Executors
import javax.xml.parsers.DocumentBuilder
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.parsers.ParserConfigurationException

@State(name = "StepikStudySettings", storages = arrayOf(Storage("stepik_study_project.xml")))
class StepikProjectManager @JvmOverloads constructor(@field:XStreamOmitField
                                                     val project: Project? = null) :
        PersistentStateComponent<Element>, DumbAware, StepikAuthManagerListener, Disposable {
    @XStreamOmitField
    private val executor = Executors.newSingleThreadExecutor()
    private var root: StudyNode<*, *>? = null
    private var selected: StudyNode<*, *>? = null
    var showHint = false
    private var createdBy: Long = 0
    var defaultLang: SupportedLanguages? = SupportedLanguages.INVALID
    var version = CURRENT_VERSION
        private set
    internal var uuid: String? = null

    private val isAdaptive: Boolean
        get() {
            if (root == null) {
                return false
            }
            val data = root!!.data
            return data != null && data.isAdaptive
        }

    val projectRoot: StudyNode<*, *>?
        get() = root

    init {
        StepikAuthManager.addListener(this)
        if (project != null) {
            Disposer.register(project, this)
        }
    }

    private fun updateAdaptiveSelected() {
        if (isAdaptive && root != null) {
            val recommendation = StudyUtils.getRecommendation(root!!)
            if (selected == null || recommendation != null && selected!!.parent !== recommendation.parent) {
                ApplicationManager.getApplication().invokeAndWait {
                    for (file in FileEditorManager.getInstance(project!!).openFiles) {
                        FileEditorManager.getInstance(project).closeFile(file)
                    }
                }
                setSelected(recommendation)
            }
        }
    }

    private fun getSelected(): StudyNode<*, *>? {
        return selected
    }

    @JvmOverloads
    fun setSelected(selected: StudyNode<*, *>?, force: Boolean = false) {
        this.selected = selected
        if (project != null) {
            val toolWindow = StudyUtils.getStudyToolWindow(project)
            if (toolWindow != null) {
                ApplicationManager.getApplication().invokeLater {
                    toolWindow.setStepNode(selected, force)
                    if (selected != null && !project.isDisposed) {
                        ProjectView.getInstance(project).refresh()
                        val file = project.baseDir.findFileByRelativePath(selected.path)
                        ProjectView.getInstance(project).select(null, file, true)
                    }
                }
            }
        }
    }

    private fun updateSelection() {
        setSelected(selected, true)
    }

    fun setRootNode(root: StudyNode<*, *>?) {
        this.root = root
    }

    override fun getState(): Element? {
        if (projectRoot == null) {
            return null
        }
        try {
            ByteArrayOutputStream().use { out ->
                xStream.toXML(this, OutputStreamWriter(out, UTF_8))
                val el = toElement(out)
                logger.info("Getting the StepikProjectManager state")

                return el
            }
        } catch (e: ParserConfigurationException) {
            logger.warn("Failed getting the StepikProjectManager state", e)
        } catch (e: IOException) {
            logger.warn("Failed getting the StepikProjectManager state", e)
        } catch (e: SAXException) {
            logger.warn("Failed getting the StepikProjectManager state", e)
        }

        return null
    }

    override fun loadState(state: Element) {
        var myState = state
        try {
            logger.info("Start load the StepikProjectManager state")
            val version = StudySerializationUtils.getVersion(myState)

            when (version) {
                1 -> {
                    myState = StudySerializationUtils.convertToSecondVersion(myState)
                    myState = StudySerializationUtils.convertToThirdVersion(myState)
                    myState = StudySerializationUtils.convertToFourthVersion(myState)
                }
                2 -> {
                    myState = StudySerializationUtils.convertToThirdVersion(myState)
                    myState = StudySerializationUtils.convertToFourthVersion(myState)
                }
                3 -> myState = StudySerializationUtils.convertToFourthVersion(myState)
            }//uncomment for future versions
            //case 4:
            //myState = StudySerializationUtils.convertToFifthVersion(myState);

            val xml = elementToXml(myState)
            xStream.fromXML(xml, this)

            this.version = CURRENT_VERSION
            refreshCourse()
            updateSelection()
            logger.info("The StepikProjectManager state loaded")
        } catch (e: XStreamException) {
            logger.warn("Failed deserialization StepikProjectManager \n${e.message}\n$project")
        } catch (e: StudyUnrecognizedFormatException) {
            logger.warn("Failed deserialization StepikProjectManager \n${e.message}\n$project")
        }

    }

    private fun refreshCourse() {
        if (project == null || root == null) {
            return
        }

        root!!.setProject(project)

        executor.execute {
            val stepikApiClient = authAndGetStepikApiClient()
            if (isAuthenticated) {
                root!!.reloadData(project, stepikApiClient)
            }
            ProgressManager.getInstance().run(object : Task.Backgroundable(project, "Synchronize Project") {
                override fun run(indicator: ProgressIndicator) {
                    if (project!!.isDisposed) {
                        return
                    }

                    repairProjectFiles(root!!)
                    repairSandbox()

                    ApplicationManager.getApplication().invokeLater {
                        VirtualFileManager.getInstance().syncRefresh()
                        setSelected(selected, false)
                    }
                }

                private fun repairSandbox() {
                    val projectDir = project!!.baseDir
                    if (projectDir != null && projectDir.findChild(EduNames.SANDBOX_DIR) == null) {
                        val application = ApplicationManager.getApplication()

                        val model = application.runReadAction(Computable {
                            ModuleManager.getInstance(project!!)
                                    .modifiableModel
                        })

                        application.invokeLater {
                            application.runWriteAction {
                                try {
                                    SandboxModuleBuilder(projectDir.path).createModule(model)
                                    model.commit()
                                } catch (e: IOException) {
                                    logger.warn("Failed repair Sandbox", e)
                                } catch (e: ConfigurationException) {
                                    logger.warn("Failed repair Sandbox", e)
                                } catch (e: JDOMException) {
                                    logger.warn("Failed repair Sandbox", e)
                                } catch (e: ModuleWithNameAlreadyExists) {
                                    logger.warn("Failed repair Sandbox", e)
                                }
                            }
                        }
                    }
                }
            })
        }
    }

    private fun repairProjectFiles(node: StudyNode<*, *>) {
        if (project != null) {
            if (node is StepNode) {
                ApplicationManager.getApplication().invokeAndWait {
                    if (project.isDisposed) {
                        return@invokeAndWait
                    }
                    ProjectFilesUtils.getOrCreateSrcDirectory(project, node, false)
                }
            }
            node.children.forEach { this.repairProjectFiles(it) }
        }
    }

    fun setCreatedBy(createdBy: Long) {
        this.createdBy = createdBy
    }

    fun getUuid(): String {
        if (uuid == null) {
            uuid = UUID.randomUUID().toString()
        }
        return uuid!!
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false

        val that = other as StepikProjectManager?

        if (showHint != that!!.showHint) return false
        if (createdBy != that.createdBy) return false
        if (version != that.version) return false
        if (if (project != null) project != that.project else that.project != null) return false
        if (if (root != null) root != that.root else that.root != null) return false

        if (defaultLang !== that.defaultLang) return false
        return if (uuid != null) uuid == that.uuid else that.uuid == null
    }

    override fun hashCode(): Int {
        var result = project?.hashCode() ?: 0
        result = 31 * result + if (root != null) root!!.hashCode() else 0
        result = 31 * result + if (showHint) 1 else 0
        result = 31 * result + (createdBy xor createdBy.ushr(32)).toInt()
        result = 31 * result + if (defaultLang != null) defaultLang!!.hashCode() else 0
        result = 31 * result + version
        result = 31 * result + if (uuid != null) uuid!!.hashCode() else 0
        return result
    }

    override fun stateChanged(oldState: StepikAuthState, newState: StepikAuthState) {
        if (newState === NOT_AUTH || newState === AUTH) {
            if (root != null) {
                root!!.resetStatus()
                if (newState === AUTH) {
                    refreshCourse()
                }
            }

            if (oldState === SHOW_DIALOG && newState === NOT_AUTH) {
                setSelected(selected, false)
            } else {
                updateSelection()
            }
        }
    }

    override fun dispose() {
        StepikAuthManager.removeListener(this)
        executor.shutdown()
    }

    companion object {
        private const val CURRENT_VERSION = 4
        private val logger = Logger.getInstance(StepikProjectManager::class.java)
        @XStreamOmitField
        val xStream: XStream = {
            XStream(DomDriver()).apply {
                alias("StepikProjectManager", StepikProjectManager::class.java)
                alias("CourseNode", CourseNode::class.java)
                alias("SectionNode", SectionNode::class.java)
                alias("LessonNode", LessonNode::class.java)
                alias("StepNode", StepNode::class.java)
                alias("Limit", Limit::class.java)
                alias("SupportedLanguages", SupportedLanguages::class.java)
                alias("VideoUrl", VideoUrl::class.java)
                alias("LinkedTreeMap", LinkedTreeMap::class.java)
                alias("Sample", Sample::class.java)
                alias("Course", Course::class.java)
                alias("Section", Section::class.java)
                alias("CompoundUnitLesson", CompoundUnitLesson::class.java)
                alias("Step", Step::class.java)
                alias("User", User::class.java)
                autodetectAnnotations(true)
                classLoader = StepikProjectManager::class.java.classLoader
                registerConverter(SupportedLanguagesConverter())
                registerConverter(SampleConverter())
                ignoreUnknownElements()
                setMode(XStream.ID_REFERENCES)
            }
        }.invoke()
        @XStreamOmitField
        private var outputter: XMLOutputter? = null
        @XStreamOmitField
        private var factory: DocumentBuilderFactory? = null
        @XStreamOmitField
        private var builder: DocumentBuilder? = null
        @XStreamOmitField
        private var domBuilder: DOMBuilder? = null

        fun getInstance(project: Project): StepikProjectManager? {
            return ServiceManager.getService(project, StepikProjectManager::class.java)
        }

        fun getProjectRoot(project: Project): StudyNode<*, *>? {
            if (project.isDisposed) {
                return null
            }
            val instance = getInstance(project) ?: return null

            return instance.projectRoot
        }

        fun isStepikProject(project: Project?): Boolean {
            if (project == null) {
                return false
            }
            val instance = getInstance(project)
            return instance?.projectRoot != null
        }

        private fun elementToXml(state: Element): String {
            if (outputter == null) {
                outputter = XMLOutputter()
            }
            return outputter!!.outputString(state.getChild(StudySerializationUtils.MAIN_ELEMENT))
        }

        @Throws(ParserConfigurationException::class, SAXException::class, IOException::class)
        private fun toElement(out: ByteArrayOutputStream): Element {
            if (factory == null) {
                factory = DocumentBuilderFactory.newInstance()
                factory!!.isValidating = false
                builder = factory!!.newDocumentBuilder()
                domBuilder = DOMBuilder()
            }
            ByteArrayInputStream(out.toByteArray()).use { `in` ->
                val doc = builder!!.parse(`in`)
                val document = domBuilder!!.build(doc)

                val root = document.rootElement
                document.removeContent(root)

                val element = Element("element")
                element.addContent(root)
                return element
            }
        }

        @JvmOverloads
        fun setSelected(project: Project, studyNode: StudyNode<*, *>?, force: Boolean = false) {
            val instance = getInstance(project)
            instance?.setSelected(studyNode, force)
        }

        fun updateSelection(project: Project) {
            val instance = getInstance(project)
            instance?.updateSelection()
        }

        fun getSelected(project: Project?): StudyNode<*, *>? {
            if (project == null) {
                return null
            }
            val instance = getInstance(project)
            return instance?.getSelected()

        }

        fun isAdaptive(project: Project): Boolean {
            val instance = getInstance(project)
            return instance != null && instance.isAdaptive
        }

        fun updateAdaptiveSelected(project: Project?) {
            if (project == null) {
                return
            }
            val instance = getInstance(project)
            instance?.updateAdaptiveSelected()
        }
    }
}
